### PR TITLE
avoid print mount recursion fix #592

### LINKED ISSUE #592: Printing recurses forever when two routers are mounted to each other
This should not fail...

```r
a <- plumber$new()
b <- plumber$new()

a$mount("B", b)
b$mount("A", a)

a
```

Maybe do something like...
```
❯❯ a
# Plumber router with 0 endpoints, 4 filters, and 1 sub-router.
# Call run() on this object to start the API.
├──[queryString]
├──[postBody]
├──[cookieParser]
├──[sharedSecret]
├──/B
│  │ # Plumber router with 0 endpoints, 4 filters, and 1 sub-router.
│  ├──[queryString]
│  ├──[postBody]
│  ├──[cookieParser]
│  ├──[sharedSecret]
│  ├──/A
│  │  │ # Plumber router with 0 endpoints, 4 filters, and 1 sub-router.
│  │  ├──[queryString]
│  │  ├──[postBody]
│  │  ├──[cookieParser]
│  │  ├──[sharedSecret]
│  │  ├──/B
│  │  │  # Circular reference detected
```

### PR DESCRIPTION
**Pull request by meztez**
Minimal reproducible example

#592
PR task list:
- [x] Update NEWS
- [x] Add tests
- [x] Update documentation with `devtools::document()`

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -3,7 +3,9 @@ plumber 1.0.0.9999 Development version
 
 ### Bug fixes
 
-* When calling `Plumber$handle()` and defining a new `PlumberEndpoint`, `...` will be checked for invalid names #677  
+* When calling `Plumber$handle()` and defining a new `PlumberEndpoint`, `...` will be checked for invalid names #677
+
+* When calling `print` on router, detect mount recursion and exit early #683
```
Comment (schloerke) [inline NEWS.md:8]:
```suggestion
* When calling `print` on router, detect mount recursion and exit early (@meztez, #683)
```

File: R/plumber.R [hunk 2]
```diff
@@ -382,6 +382,15 @@ Plumber <- R6Class(
     #' @param ... additional arguments for recursive calls
     #' @return A terminal friendly representation of a plumber router.
     print = function(prefix="", topLevel=TRUE, ...){
+
+      # Avoid printing recursion (mount on mount on mount on ...)
+      on.exit({private$printing = NULL}, add = TRUE)
+      private$printing = sum(private$printing, ifelse(topLevel, 0L, 1L))
+      if (isTRUE(private$printing == 2L)) {
+        cat(prefix, "# Circular reference detected", "\n", sep="")
```
Comment (schloerke) [inline R/plumber.R:390]:
```suggestion
        cat(prefix, "# Circular Plumber reference detected", "\n", sep="")
```

File: tests/testthat/test-plumber-print.R [hunk 3]
```diff
@@ -79,3 +79,20 @@ test_that("prints correctly", {
...
+
+  a$mount("B", b)
+  b$mount("A", a)
+
+  printed <- capture.output(print(a))
+  printed2 <- capture.output(print(a))
+  printed3 <- capture.output(print(b))
+
+  expect_match(printed, "Circular reference detected", all = FALSE)
+  expect_match(printed3, "Circular reference detected", all = FALSE)
```
Comment (schloerke) [inline tests/testthat/test-plumber-print.R:95]:
```suggestion
  expect_match(printed, "Circular Plumber reference detected", all = FALSE)
  expect_match(printed3, "Circular Plumber reference detected", all = FALSE)
```

**Review by schloerke — CHANGES_REQUESTED** [review]

**Review by schloerke — APPROVED** [review]

**Review by cpsievert — APPROVED** [review]
If you want, it might be nice to have a different color for the circular ref message

**Comment by meztez** [comment]
> If you want, it might be nice to have a different color for the circular ref message

Any color you fancy?

**Comment by meztez** [comment]
currentyl using  `printing`, maybe use flag facility when implemented.

**Comment by schloerke** [comment]
Closing in favor of #738

**Pull request closed without merging**